src/modules/ekf2/EKF/mag_3d_control.cpp

```diff
@@ -0,0 +1,215 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mag_3d_control.cpp
+ * Control functions for ekf mag 3D fusion
+ */
+
+#include "ekf.h"
+
+void Ekf::controlMag3DFusion(const magSample &mag_sample, const bool common_starting_conditions_passing,
+			     estimator_aid_source3d_s &aid_src)
+{
+	static constexpr const char *AID_SRC_NAME = "mag";
+
+	resetEstimatorAidStatus(aid_src);
+
+	const bool wmm_updated = (_wmm_gps_time_last_set > aid_src.time_last_fuse);
+
+	// determine if we should use mag fusion
+	bool continuing_conditions_passing = (_params.mag_fusion_type != MagFuseType::NONE)
+					     && _control_status.flags.tilt_align
+					     && (_control_status.flags.yaw_align || (!_control_status.flags.ev_yaw && !_control_status.flags.yaw_align))
+					     && (wmm_updated || isRecent(_time_last_mov_3d_mag_suitable, (uint64_t)3e6))
+					     && mag_sample.mag.longerThan(0.f)
+					     && mag_sample.mag.isAllFinite();
+
+	const bool starting_conditions_passing = common_starting_conditions_passing
+			&& continuing_conditions_passing;
+
+	// For the first few seconds after in-flight alignment we allow the magnetic field state estimates to stabilise
+	// before they are used to constrain heading drift
+	_control_status.flags.mag_3D = (_params.mag_fusion_type == MagFuseType::AUTO)
+				       && _control_status.flags.mag
+				       && _control_status.flags.mag_aligned_in_flight && isTimedOut(_flt_mag_align_start_time, (uint64_t)5e6)
+				       && !_control_status.flags.mag_fault
+				       && !magFieldStrengthDisturbed(mag_sample.mag - _state.mag_B)
+				       && getMagBiasVariance().longerThan(0.f) && !getMagBiasVariance().longerThan(sq(0.02f))
+				       && !_control_status.flags.ev_yaw
+				       && !_control_status.flags.gps_yaw;
+
+	// TODO: allow clearing mag_fault if mag_3d is good?
+
+	if (_control_status.flags.mag_3D && !_control_status_prev.flags.mag_3D) {
+		ECL_INFO("starting mag 3D fusion");
+
+	} else if (!_control_status.flags.mag_3D && _control_status_prev.flags.mag_3D) {
+		ECL_INFO("stopping mag 3D fusion");
+	}
+
+	// if we are using 3-axis magnetometer fusion, but without external NE aiding,
+	// then the declination must be fused as an observation to prevent long term heading drift
+	// fusing declination when gps aiding is available is optional, but recommended to prevent
+	// problem if the vehicle is static for extended periods of time
+	const bool mag_decl_user_selected = (_params.mag_declination_source & GeoDeclinationMask::FUSE_DECL);
+	const bool not_using_ne_aiding = !_control_status.flags.gps;
+	_control_status.flags.mag_dec = (_control_status.flags.mag && (not_using_ne_aiding || mag_decl_user_selected));
+
+	if (_control_status.flags.mag) {
+		aid_src.timestamp_sample = mag_sample.time_us;
+		aid_src.fusion_enabled = true;
```

Comment:
but if the `continuing_conditions_passing == false`, do we still want to have `fusion_enabled = true`?

Reply:
Yes that's not quite right, we need to revamp the aid_src population so it's correct and useful on start/reset/stop cycles.

For now I'll move setting fusion_enabled to the end.